node/src/coordinator.rs

```diff
@@ -736,3 +683,131 @@ impl Drop for ReportCurrentJobGuard {
+        stop_running(
+            new_state,
+            current_running_epoch_id,
+            current_participant_set.clone(),
+        )
+    })
+}
+
+/// returns true if one of the following occurs:
+///     - epoch id changed
+///     - resharing concludes
+///     - key event receiver closes the channel.
+fn stop_resharing(
+    new_state: &ContractState,
+    current_resharing_epoch_id: EpochId,
+    key_event_sender: &tokio::sync::watch::Sender<ContractKeyEventInstance>,
+) -> bool {
+    match new_state {
+        ContractState::Running(new_state) => {
+            let Some(new_resharing_state) = &new_state.resharing_state else {
+                tracing::info!("Concluded resharing state.");
+                return true;
+            };
+
+            if new_resharing_state.key_event.id.epoch_id != current_resharing_epoch_id {
+                tracing::info!("Epoch changed. We exit resharing state.");
+                return true;
+            }
+
+            if key_event_sender
+                .send(new_resharing_state.key_event.clone())
+                .is_err()
+            {
+                tracing::info!("Key event receiver closed.");
+                return true;
+            }
+
+            false
+        }
+        _ => true,
+    }
+}
+
+fn make_resharing_stop_fn(
+    resharing_state: ContractResharingState,
+) -> (watch::Receiver<ContractKeyEventInstance>, StopFn) {
+    let (key_event_sender, key_event_receiver) = watch::channel(resharing_state.key_event.clone());
+    let current_resharing_epoch_id = resharing_state.key_event.id.epoch_id;
+    let stop_fn = Box::new(move |new_state: &ContractState| {
+        stop_resharing(new_state, current_resharing_epoch_id, &key_event_sender)
+    });
+    (key_event_receiver, stop_fn)
+}
+
+fn stop_initializing(
+    new_state: &ContractState,
+    current_epoch_id: EpochId,
+    key_event_sender: &tokio::sync::watch::Sender<ContractKeyEventInstance>,
+) -> bool {
+    match new_state {
+        ContractState::Initializing(new_state) => {
+            if new_state.key_event.id.epoch_id != current_epoch_id {
+                tracing::info!("Epoch id changed");
+                return true;
+            }
+            if key_event_sender.send(new_state.key_event.clone()).is_err() {
+                tracing::info!("Key event receiver closed");
+                return true;
+            }
+            false
+        }
+        _ => {
+            tracing::info!("Protocol State changed.");
+            true
+        }
+    }
+}
+
+fn make_initializing_stop_fn(
+    key_event: ContractKeyEventInstance,
+) -> (watch::Receiver<ContractKeyEventInstance>, StopFn) {
+    let (key_event_sender, key_event_receiver) = watch::channel(key_event.clone());
+    let key_event_sender = key_event_sender.clone();
+    (
+        key_event_receiver,
+        Box::new(move |new_state| {
+            stop_initializing(new_state, key_event.id.epoch_id, &key_event_sender)
+        }),
+    )
+}
```

Comment:
Just so that I'm not missing anything important, there's no logic change here right? Except for the extra log line if the key event receiver is closed. Beyond that, this seems like a pure (and nice) refactor - correct?

Reply:
Exactly, I didn't intend to change any logic here, just a refactor.